wgsl/index.bs

```diff
@@ -7293,13 +7291,330 @@ See [[#expressions]].
+- Create two new nodes, named "Result" and "CF_after"
+- If the control-flow at the start of the function was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to the last CF_i
+- Otherwise add an edge from CF_after to the last CF_i
+- If the function was tagged ControlFlowCannotBeUniformAfter, then add an edge from CF_after to CannotBeUniform
+- Else if it was tagged ReturnValueCannotBeUniform, then add an edge from Result to CannotBeUniform
+- Add an edge from Result to CF_after
+- For each argument *i*:
+    - If the corresponding parameter was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to arg_i
+    - Else if it was tagged MustBeUniformForControlFlowAfter, then add an edge from CF_after to arg_i
+    - Else if it was tagged MustBeUniformForReturnValue, then add an edge from Result to arg_i
+
+Note: Notice that this rule only requires adding a number of edges bounded by 3 + the number of parameters of the functions, independently of how complex the implementation of the function might be. This is key to the linear complexity of the overall algorithm.
+
+Most built-in functions have tags of:
+- for the control-flow in which they can be called: MustBeUniformForControlFlowAfter
+- for the function as a whole: NoRestriction
+- for each parameter: MustBeUniformForReturnValue
+
+Here is the list of exceptions:
+- All functions in [[#sync-builtin-functions]] are AlwaysMustBeUniform
+- All functions in [[#derivative-builtin-functions]], [[#texturesample]], [[#texturesamplebias]], and [[#texturesamplecompare]] are AlwaysMustBeUniform and ReturnValueCannotBeUniform
+
+### Uniformity rules for expressions ### {#uniformity-expressions}
+
+The rules for analyzing expressions take as argument both the expression itself and the node corresponding to control flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A node corresponding to control flow at the exit of it
+* A node corresponding to its value
+* A set of new nodes and edges to add to the graph
+
+Like for statements, if the set of behaviors for an expression is {Next}, then the node corresponding to control flow at the exit of it is replaced by *CF* (the control flow at the beginning of it), overriding what the rules below say about the resulting control-flow node.
+
+<table class='data'>
+  <caption>Uniformity rules for expressions (in normal rvalue position)</caption>
+  <thead>
+    <tr><th>Expression<th>New nodes<th>Recursive analyses<th>Resulting control flow node (if behavior is not {Next}), value node<th>New edges
+  </thead>
+  <tr><td class="nowrap">e1 || e2
+      <td>
+      <td class="nowrap">(*CF*, *e1*) => (*CF1*, *V1*)<br>
+          (*V1*, *e2*) => (*CF2*, *V2*)
```

Comment:
Should there be an edge from CF2 -> CF1 ?

The idea is:
- evaluation of e1 may cause non-uniform CF
- evaluation of e2 may require uniform CF

Reply:
I think this is a  non-issue, with the following reasoning:
V1  uniform implies CF1 uniform, so it's like there is already an edge from V1 to CF1.

Reply:
Correct. More precisely, there is a lemma that there must be a path from V1 to CF1, since (CF1, V1) was the result of applying the analysis to an expression.